In [11]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, recall_score, confusion_matrix, classification_report, roc_curve
from sklearn import svm
from sklearn.preprocessing import StandardScaler

In [12]:
df=pd.read_csv('data\PreprocessedBankChurners.csv')
df.head()

Attrition_Flag  Customer_Age  Dependent_count  Months_on_book  \
0               0            50                3              39   
1               0            60                5              44   
2               0            60                3              36   
3               0            50                4              34   
4               0            50                3              21   

   Total_Relationship_Count  Months_Inactive_12_mon  Contacts_Count_12_mon  \
0                         5                       1                      3   
1                         6                       1                      2   
2                         4                       1                      0   
3                         3                       4                      1   
4                         5                       1                      0   

   Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  ...  \
0       12691.0                  777          11914.0  ...   
1        8256.0                  864           7392.0  ...   
2        3418.0                    0           3418.0  ...   
3        3313.0                 2517            796.0  ...   
4        4716.0                    0           4716.0  ...   

   Education_Level_Doctorate  Education_Level_Graduate  \
0                          0                         0   
1                          0                         1   
2                          0                         1   
3                          0                         0   
4                          0                         0   

   Education_Level_High School  Education_Level_Post-Graduate  \
0                            1                              0   
1                            0                              0   
2                            0                              0   
3                            1                              0   
4                            0                              0   

   Education_Level_Uneducated  Education_Level_Unknown  \
0                           0                        0   
1                           0                        0   
2                           0                        0   
3                           0                        0   
4                           1                        0   

   Marital_Status_Divorced  Marital_Status_Married  Marital_Status_Single  \
0                        0                       1                      0   
1                        0                       0                      1   
2                        0                       1                      0   
3                        0                       0                      0   
4                        0                       1                      0   

   Marital_Status_Unknown  
0                       0  
1                       0  
2                       0  
3                       1  
4                       0  

[5 rows x 38 columns]

In [13]:
X = df.iloc[:, df.columns != 'Attrition_Flag']
y = df['Attrition_Flag']
X_train, X_test, y_train,y_test = train_test_split(X, y, test_size=0.25, random_state=21)

In [14]:
svc = svm.SVC()
svc.get_params().keys()

dict_keys(['C', 'break_ties', 'cache_size', 'class_weight', 'coef0', 'decision_function_shape', 'degree', 'gamma', 'kernel', 'max_iter', 'probability', 'random_state', 'shrinking', 'tol', 'verbose'])

In [21]:
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [22]:
param_grid = {    
    'C': [1,10, 100, 1000]} 
svc = svm.SVC()

svc_cv= GridSearchCV(svc,param_grid,cv=5)
svc_cv.fit(X_train_scaled,y_train)

GridSearchCV(cv=5, estimator=SVC(), param_grid={'C': [1, 10, 100, 1000]})

In [23]:
svc_cv.best_params_

{'C': 10}

In [26]:
svc_best_param = svm.SVC(C=10, random_state=42)
svc_best_param.fit(X_train_scaled, y_train)
y_pred = svc_best_param.predict(X_test_scaled)

In [27]:
print(confusion_matrix(y_test, y_pred))

[[2030   80]
 [ 137  285]]


In [28]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      0.96      0.95      2110
           1       0.78      0.68      0.72       422

    accuracy                           0.91      2532
   macro avg       0.86      0.82      0.84      2532
weighted avg       0.91      0.91      0.91      2532



Recall and f1-score is even better this time. Its 68% for churned customers.